File: Embarcadero/Berlin/CPP/HttpDownload/FDownloadDemo.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#include <fmx.h>
#pragma hdrstop

#include "FDownloadDemo.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.fmx"
//---------------------------------------------------------------------------
TForm2 *Form2;
//---------------------------------------------------------------------------
__fastcall TForm2::TForm2(TComponent* Owner)
	: TForm(Owner)
{
	FAllowClossing = true;
}

//---------------------------------------------------------------------------
void __fastcall TForm2::BStartDownloadClick(TObject *Sender)
{
	((TButton *)Sender)->Enabled = false;
	FAllowClossing = false;
	TThread::CreateAnonymousThread(AnonymousLambda(&this->SampleDownload))->Start();
}
//---------------------------------------------------------------------------
void __fastcall TForm2::FormCreate(TObject *Sender)
{
	ButtonCancelArray[0] = Button1;
	ButtonCancelArray[1] = Button2;
	ButtonCancelArray[2] = Button3;
	ButtonCancelArray[3] = Button4;
	ProgressBarArray[0] = ProgressBarPart1;
	ProgressBarArray[1] = ProgressBarPart2;
	ProgressBarArray[2] = ProgressBarPart3;
	ProgressBarArray[3] = ProgressBarPart4;
	LabelProgressArray[0] = Label1;
	LabelProgressArray[1] = Label2;
	LabelProgressArray[2] = Label3;
	LabelProgressArray[3] = Label4;
}
//---------------------------------------------------------------------------
void TForm2::ResetProgress(int Index, int MaxValue)
{
	ProgressBarArray[Index]->Max =  MaxValue;
	ProgressBarArray[Index]->Min = 0;
	ProgressBarArray[Index]->Value = 0;
	ButtonCancelArray[Index]->Enabled = True;
	LabelProgressArray[Index]->Text = "0 KB/s";
}
//---------------------------------------------------------------------------
void TForm2::ShowGlobalString(UnicodeString AnString)
{
	LabelGlobalSpeed->Text = AnString;
}
//---------------------------------------------------------------------------
void TForm2::AddLineToMemo(UnicodeString AnString)
{
	Memo1->Lines->Add(AnString);
}
//---------------------------------------------------------------------------
void TForm2::EnableStartButton()
{
	BStartDownload->Enabled = true;
}
//---------------------------------------------------------------------------
void TForm2::SampleDownload(void)
{
	const int NumThreads = 4;
	THTTPClient *LClient;
	LClient = THTTPClient::Create();

	try
	{
		String LFileName = EditFileName->Text;
		String LPath = System::Ioutils::TPath::Combine(System::Ioutils::TPath::GetDocumentsPath(), LFileName);
		System::Classes::TThread::Synchronize(NULL, SyncLambda<UnicodeString>(&this->AddLineToMemo, "File location = " + LPath));

		String URL = EditURL->Text;
		System::Classes::TThread::Synchronize(NULL, SyncLambda<UnicodeString>(&this->AddLineToMemo, "Downloading " + URL + " ..."));

		if (LClient->CheckDownloadResume(URL) == true)
		{
			_di_IHTTPResponse LResponse = LClient->Head(URL);

			// Get space for the file that is going to be dowloaded
			__int64 LSize = LResponse->ContentLength;
			System::Classes::TFileStream *STFile = new System::Classes::TFileStream(LPath, fmCreate);
			try
			{
				STFile->Size = LSize;
			}
			__finally
			{
				STFile->Free();
			};


			// Split the file in four blocks
			__int64 LFragSize = LSize / NumThreads;
			__int64 LStart = 0;
			__int64 LEnd = LStart + LFragSize;

			DynamicArray<TDownloadThread *> LDownloadThreads;
			LDownloadThreads.Length = NumThreads;
			for (int i = 0; i < NumThreads; i++)
			{
				if (FClosingForm)
					break;
				// Create the Thread
				LDownloadThreads[i] = new TDownloadThread(URL, LPath, i, LStart, LEnd);
				LDownloadThreads[i]->OnThreadData = ReceiveThreadDataEvent;

				if (LEnd >= LSize)
					LEnd = LSize;

				System::Classes::TThread::Synchronize(NULL, SyncLambda<int, int>(&this->ResetProgress, i,
					LEnd >= LSize ? (LFragSize - (LEnd - LSize)) : (LFragSize)));

				// Update Start and End Values
				LStart = LStart + LFragSize;
				LEnd = LStart + LFragSize;
			}


			// Start the download process
			__int64 LStartTime = TThread::GetTickCount();
			for (int i = 0; i < NumThreads; i++)
				LDownloadThreads[i]->Start();

			// Wait until all threads finish
			for (System::Boolean LFinished = false; LFinished == false && FClosingForm == false; )
			{
				LFinished = True;
				for (int i = 0; i < NumThreads; i++)
				  LFinished = LFinished && LDownloadThreads[i]->Finished;
			}

			// Calculate elapsed time and Average Speed;
			__int64 LEndTime = TThread::GetTickCount() - LStartTime;
			TVarRec data[1];
			data[0] = (((LSize*1000) / LEndTime) / 1024);

			System::Classes::TThread::Synchronize(NULL, SyncLambda<UnicodeString>(&this->ShowGlobalString, Format("Global Speed: %d KB/s", data, 0)));

			// Cleanup Threads
			for (int i = 0; i < NumThreads; i++)
				delete LDownloadThreads[i];
		}
		else
		{
			System::Classes::TThread::Synchronize(NULL, SyncLambda<UnicodeString>(&this->AddLineToMemo, "Server has NOT resume download feature"));
		}
	}
	__finally
	{
		LClient->Free();
		TThread::Synchronize(NULL, SyncLambda(&this->EnableStartButton));
		FAllowClossing = true;
	}
}
//---------------------------------------------------------------------------
void TForm2::UpdateProgress(int AThreadIndex, int AValue, int ASpeed, bool* Abort)
{
	*Abort = !ButtonCancelArray[AThreadIndex]->Enabled;
	ProgressBarArray[AThreadIndex]->Value = AValue;
	TVarRec LVarRec[1];
	LVarRec[0] = (ASpeed / 1024);
	LabelProgressArray[AThreadIndex]->Text = Format("%d KB/s", LVarRec, 0);
}
//---------------------------------------------------------------------------
void __fastcall TForm2::ReceiveThreadDataEvent(const TObject *Sender, int ThreadNo, int ASpeed, __int64 AContentLength,
	__int64 AReadCount, System::Boolean &Abort)
{
	Abort = FClosingForm;
	if (Abort == false) {
		bool abort;
		System::Classes::TThread::Synchronize(NULL, SyncLambda<int, int, int, bool*>(&this->UpdateProgress, ThreadNo, AReadCount, ASpeed, &abort));
		Abort = abort;
	}
}
//---------------------------------------------------------------------------
__fastcall TDownloadThread::TDownloadThread(UnicodeString &AURL, UnicodeString &AFileName, int AThreadNo,
	__int64 StartPos, __int64 EndPos): TThread(true), FURL(AURL), FFileName(AFileName), FThreadNo(AThreadNo),
	FStartPos(StartPos), FEndPos(EndPos)
{
}
//---------------------------------------------------------------------------
void __fastcall TDownloadThread::Execute(void)
{
	THTTPClient *LHttpClient = THTTPClient::Create();
	try
	{
		LHttpClient->OnReceiveData = ReceiveDataEvent;
		TFileStream *LStream = new TFileStream(FFileName, fmOpenWrite | fmShareDenyNone);
		try
		{
		  LStream->Seek(FStartPos, TSeekOrigin::soBeginning);
		  FTimeStart = GetTickCount();
		  _di_IHTTPResponse LResponse = LHttpClient->GetRange(FURL, FStartPos, FEndPos, LStream);
		}
		__finally
		{
		  delete LStream;
		}
	}
	__finally
	{
		LHttpClient->Free();
	}
}
//---------------------------------------------------------------------------
void __fastcall TDownloadThread::ReceiveDataEvent(System::TObject* const Sender, __int64 AContentLength, __int64 AReadCount,
	System::Boolean &Abort)
{
	if (FOnThreadData != NULL)
	{
		unsigned int LTime = GetTickCount() - FTimeStart;
		__int64 LSpeed = (AReadCount * 1000) / LTime;
		FOnThreadData(Sender, FThreadNo, LSpeed, AContentLength, AReadCount, Abort);
	}
}

//---------------------------------------------------------------------------

void __fastcall TForm2::ButtonStopClick(TObject *Sender)
{
	((TButton *)Sender)->Enabled = false;
}
//---------------------------------------------------------------------------

void __fastcall TForm2::FormClose(TObject *Sender, TCloseAction &Action)
{
	Memo1->Lines->Add("Cancelling the download process may take some time.");
	Application->ProcessMessages();


	FClosingForm = true;
	for (int i = 0; i < 4; i++)
		ButtonCancelArray[i]->Enabled = false;

	while (FAllowClossing == false)
		Application->ProcessMessages();
}
//---------------------------------------------------------------------------

File: Embarcadero/Rio/Object Pascal/HttpDownload/FDownloadDemo.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit FDownloadDemo;

interface

uses
  System.SysUtils, System.Types, System.UITypes, System.Classes, System.Variants,
  FMX.Types, FMX.Controls, FMX.Forms, FMX.Graphics, FMX.Dialogs, System.Net.URLClient, System.Net.HttpClient,
  System.Net.HttpClientComponent, FMX.StdCtrls, FMX.Edit, FMX.Controls.Presentation, FMX.ScrollBox, FMX.Memo,
  System.ImageList, FMX.ImgList;

type
  TDownloadThreadDataEvent = procedure(const Sender: TObject; ThreadNo, ASpeed: Integer; AContentLength: Int64; AReadCount: Int64; var Abort: Boolean) of object;
  TDownloadThread = class(TThread)
  private
    FOnThreadData: TDownloadThreadDataEvent;

  protected
    FURL, FFileName: string;
    FStartPoint, FEndPoint: Int64;
    FThreadNo: Integer;
    FTimeStart: Cardinal;

    procedure ReceiveDataEvent(const Sender: TObject; AContentLength: Int64; AReadCount: Int64; var Abort: Boolean);
  public
    constructor Create(const URL, FileName: string; ThreadNo: Integer; StartPoint, EndPoint: Int64);
    destructor Destroy; override;
    procedure Execute; override;

    property OnThreadData: TDownloadThreadDataEvent write FOnThreadData;
  end;

  TFormDownloadDemo = class(TForm)
    PanelTop: TPanel;
    PanelCenter: TPanel;
    LabelFile: TLabel;
    EditFileName: TEdit;
    BStartDownload: TButton;
    Memo1: TMemo;
    ProgressBarPart1: TProgressBar;
    Label1: TLabel;
    Button1: TButton;
    ImageList1: TImageList;
    LabelURL: TLabel;
    EditURL: TEdit;
    LabelGlobalSpeed: TLabel;
    Panel1: TPanel;
    Panel2: TPanel;
    Button4: TButton;
    Label4: TLabel;
    ProgressBarPart4: TProgressBar;
    Panel3: TPanel;
    Button3: TButton;
    Label3: TLabel;
    ProgressBarPart2: TProgressBar;
    Panel4: TPanel;
    Button2: TButton;
    Label2: TLabel;
    ProgressBarPart3: TProgressBar;
    Panel5: TPanel;
    procedure BStartDownloadClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure ButtonCancelClick(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
  private
    { Private declarations }
    ButtonCancelArray: array [0..3] of TButton;
    ProgressBarArray: array [0..3] of TProgressBar;
    LabelProgressArray: array [0..3] of TLabel;
    [volatile] FAllowFormClose: Boolean;
    [volatile] FClosingForm: Boolean;
    procedure ReceiveThreadDataEvent(const Sender: TObject; ThreadNo, ASpeed: Integer; AContentLength: Int64; AReadCount: Int64; var Abort: Boolean);
  public
  const
    NumThreads: Integer = 4;
  public
    { Public declarations }
    procedure SampleDownload;
  end;

var
  FormDownloadDemo: TFormDownloadDemo;

implementation

{$R *.fmx}

uses
  System.IOUtils;

procedure TFormDownloadDemo.BStartDownloadClick(Sender: TObject);
begin
  (Sender as TButton).Enabled := False;
  FAllowFormClose := False;
  TThread.CreateAnonymousThread(procedure
  begin
    SampleDownload;
  end).Start;
end;

procedure TFormDownloadDemo.ButtonCancelClick(Sender: TObject);
begin
  (Sender as TButton).Enabled := False;
end;

procedure TFormDownloadDemo.FormClose(Sender: TObject; var Action: TCloseAction);
var
  I: Integer;
begin
	Memo1.Lines.Add('Cancelling the download process may take some time.');
	Application.ProcessMessages;

  FClosingForm := True;

  for I := Low(ButtonCancelArray) to High(ButtonCancelArray) do
    ButtonCancelArray[I].Enabled := False;

  while not FAllowFormClose do
  begin
    Application.ProcessMessages;
    Sleep(1);
  end;
end;

procedure TFormDownloadDemo.FormCreate(Sender: TObject);
begin
  FAllowFormClose := True;
  ButtonCancelArray[0] := Button1;
  ButtonCancelArray[1] := Button2;
  ButtonCancelArray[2] := Button3;
  ButtonCancelArray[3] := Button4;
  ProgressBarArray[0] := ProgressBarPart1;
  ProgressBarArray[1] := ProgressBarPart2;
  ProgressBarArray[2] := ProgressBarPart3;
  ProgressBarArray[3] := ProgressBarPart4;
  LabelProgressArray[0] := Label1;
  LabelProgressArray[1] := Label2;
  LabelProgressArray[2] := Label3;
  LabelProgressArray[3] := Label4;
end;

procedure TFormDownloadDemo.ReceiveThreadDataEvent(const Sender: TObject; ThreadNo: Integer; ASpeed: Integer; AContentLength, AReadCount: Int64;
  var Abort: Boolean);
var
  LCad: string;
  LCancel: Boolean;
  LSpeed: Integer;
begin
  LCancel := Abort or FClosingForm;
  if not LCancel then
    TThread.Synchronize(nil,
      procedure
      begin
        LCancel := not ButtonCancelArray[ThreadNo].Enabled;

        ProgressBarArray[ThreadNo].Value := AReadCount;
        LabelProgressArray[ThreadNo].Text := Format('%d KB/s', [ASpeed div 1024]);
      end);
  Abort := LCancel;
end;


procedure TFormDownloadDemo.SampleDownload;
var
  LClient: THTTPClient;
  URL: string;
  LResponse: IHTTPResponse;
  StFile: TFileStream;
  LFileName: string;
  LStart, LEnd, LSize, LFragSize: Int64;
  I: Integer;
  LDownloadThreads: array of TDownloadThread;
  LFinished: Boolean;
  LStartTime, LEndTime: Cardinal;
begin
  LClient := THTTPClient.Create;
  LFileName := TPath.Combine(TPath.GetDocumentsPath, EditFileName.Text);

  TThread.Synchronize(nil, procedure
  begin
    Memo1.Lines.Add('File location = ' + LFileName);
    Memo1.Lines.Add('Downloading ' + URL + ' ...');
    Application.ProcessMessages;
  end);
  try
    URL := EditURL.Text;
    if LClient.CheckDownloadResume(URL) then
    begin
      LResponse := LClient.Head(URL);

      // Get space for the file that is going to be dowloaded
      LSize := LResponse.ContentLength;
      StFile := TFileStream.Create(LFileName, fmCreate);
      try
        STFile.Size := LSize;
      finally
        STFile.Free;
      end;

      // Split the file in four blocks
      LFragSize := LSize div NumThreads;
      LStart := 0;
      LEnd := LStart + LFragSize;

      SetLength(LDownloadThreads, NumThreads);
      for I := 0 to NumThreads - 1 do
      begin
        if FClosingForm then
          Break;

        // Create the Thread
        LDownloadThreads[I] := TDownloadThread.Create(URL, LFileName, I, LStart, LEnd);
        LDownloadThreads[I].OnThreadData := ReceiveThreadDataEvent;

        TThread.Synchronize(nil, procedure
        begin
          // Adjust the ProgressBar Max Value
          if LEnd >= LSize then
          begin
            ProgressBarArray[I].Max := LFragSize - (LEnd - LSize);
            LEnd := LSize;
          end
          else
            ProgressBarArray[I].Max := LFragSize;
          ProgressBarArray[I].Min := 0;
          ProgressBarArray[I].Value := 0;

          ButtonCancelArray[I].Enabled := True;
          LabelProgressArray[I].Text := '0 KB/s';
        end);

        // Update Start and End Values
        LStart := LStart + LFragSize;
        LEnd := LStart + LFragSize;
      end;

      // Start the download process
      LStartTime := TThread.GetTickCount;
      for I := 0 to NumThreads - 1 do
        LDownloadThreads[I].Start;

      // Wait until all threads finish
      LFinished := False;
      while not LFinished and not FClosingForm do
      begin
        LFinished := True;
        for I := 0 to NumThreads - 1 do
          LFinished := LFinished and LDownloadThreads[I].Finished;
      end;

      LEndTime := TThread.GetTickCount - LStartTime;
      TThread.Synchronize(nil, procedure
      begin
        LabelGlobalSpeed.Text := Format('Global Speed: %d KB/s', [((LSize*1000) div LEndTime) div 1024]);
      end);

      // Cleanup Threads
      for I := 0 to NumThreads - 1 do
        LDownloadThreads[I].Free;

    end
    else
    begin
      TThread.Synchronize(nil, procedure
      begin
        Memo1.Lines.Add('Server has NOT resume download feature');
      end);
    end;
  finally
    LClient.Free;
    TThread.Synchronize(nil, procedure
    begin
     BStartDownload.Enabled := True;
    end);
    FAllowFormClose := True;

  end;
end;

{ TDownloadThread }

constructor TDownloadThread.Create(const URL, FileName: string; ThreadNo: Integer; StartPoint, EndPoint: Int64);
begin
  inherited Create(True);
  FURL := URL;
  FFileName := FileName;
  FThreadNo := ThreadNo;
  FStartPoint := StartPoint;
  FEndPoint := EndPoint;
end;

destructor TDownloadThread.Destroy;
begin
  inherited;
end;

procedure TDownloadThread.Execute;
var
  LResponse: IHTTPResponse;
  LStream: TFileStream;
  LHttpClient: THTTPClient;
begin
  inherited;
  LHttpClient := THTTPClient.Create;
  try
    LHttpClient.OnReceiveData := ReceiveDataEvent;
    LStream := TFileStream.Create(FFileName, fmOpenWrite or fmShareDenyNone);
    try
      FTimeStart := GetTickCount;
      LStream.Seek(FStartPoint, TSeekOrigin.soBeginning);
      LResponse := LHttpClient.GetRange(FURL, FStartPoint, FEndPoint, LStream);
    finally
      LStream.Free;
    end;
  finally
    LHttpClient.Free;
  end;
end;

procedure TDownloadThread.ReceiveDataEvent(const Sender: TObject; AContentLength, AReadCount: Int64;
  var Abort: Boolean);
var
  LTime: Cardinal;
  LSpeed: Integer;
begin
  if Assigned(FOnThreadData) then
  begin
    LTime := GetTickCount - FTimeStart;
    if AReadCount = 0 then
      LSpeed := 0
    else
      LSpeed := (AReadCount * 1000) div LTime;

    FOnThreadData(Sender, FThreadNo, LSpeed, AContentLength, AReadCount, Abort);
  end;
end;

end.
